### ISSUE TITLE
Retain EventPoller in disruptor core

### ISSUE THREAD

**Issue by A248** [LMAX-Exchange/disruptor#367]
The EventPoller is very useful in certain cases, particularly where the consuming thread has other responsibilities at the same time. For example, a UI thread or a game thread. Some other users agree: https://stackoverflow.com/questions/54157804/pollable-disruptor

I would ask that EventPoller be kept inside the disruptor and not removed.

Also, unlike with the WorkerPool, it is not as simple to recreate the functionality of the class otherwise, whereas usages of WorkerPool can be changed to use striped event handlers. Moreover, EventPoller is far less of a maintenance burden.

**Comment by grumpyjames** [comment]
tl;dr Ok, we'll keep it - we would _eagerly_ accept a PR adding javadoc for this component to help resolve #365 

**Longer version:**

Things I really need to do: review my Github notifications settings; totally didn't realise this was on my to do list. Please accept my apologies for the unbelievably slow reply.

This decision came from an exploration of https://github.com/LMAX-Exchange/disruptor/issues/365 , and a realisation that this isn't a component we use internally at LMAX. Typically we want to minimise the surface area of the library, and so my first instinct was "can we evict it?"

I've taken a second look at our internal `RingBuffer` usage now, and what I am finding is that while we don't use `EventPoller`, we sure seem to write code that reimplements it quite a lot, potentially incorrectly. I think this probably means it should stay.

Could I tempt you into providing some javadoc around this component, if you have familiarity with it?

**Comment by A248** [comment]
Sure, I'll happily make a PR for this. There's no rush; it is OSS after all.

As I am also so inclined, I totally understand the need to minimize API surface or else make what's exposed high-quality API. The EventPoller is still marked as experimental. Are you looking for more tests before you declare this a stable feature?

**Comment by grumpyjames** [comment]
Thanks for putting that pull request together!

Re: Taking EventPoller from experimental to stable.

There certainly could be "more tests" component to that - although coming to it cold I'm not sure what the first additional test I'd write is. If you look at `EventPollerTest` and think "well it's unforgivable to have not explored this and this code path" we would, again, eagerly accept improvements.

The second half is trickier; it's about whether `EventPoller` is getting used regularly without issue. We make heavy use of the rest of the library internally at LMAX; that's what gives us the confidence that we have something we're happy to call stable. We should probably have a look around the open source ecosystem to see if `EventPoller` has any adherents that are making use of it successfully. Or, alternatively, we should replace one or two of the lookalikes in the LMAX codebase with it and see what happens...

**Comment by ennerf** [comment]
@grumpyjames in case you are looking for feedback.

The EventPoller is very useful for GUI applications where all the UI updates need to happen on a specific thread that is controlled by an external framework. For our JavaFX applications we use `EventPoller` in combination with `AnimationTimer` that gets called once per frame. This way the events are automatically batched in frame rate based chunks, we can gather data from multiple sources coming in on different threads, and we can get rid of pretty much all other forms of synchronization. It looks something like this:

```java
new AnimationTimer() {
@Override
public void handle(long now) {
  while(poller.poll(handler) == PollState.PROCESSING) {
  }
}.start();
```

Overall, we are very happy with how it works, but there are definitely a few pain points in the current API:

* Lack of `LifecycleAware` support. This would be very nice for starting/stopping the AnimationTimer. The start requires an order (1. add poller, 2. start disruptor, 3. start polling),  and many instances don't live for the full duration of the application.
* Lack of `BatchAware` support
* The different interface (boolean return) can be annoying when moving an EventHandler from a background thread to the JavaFX thread. As far as I know we have yet to come across a use case where a handler needed to exit a batch early. Our use cases so far have either required only the last event (e.g. live streaming video), or an aggregate of all events (e.g. plotting live time series data).

**Comment by grumpyjames** [comment]
This has been open for ages because I _still_ need to take a bit of time to see if we should be using EventPoller internally, and having this issue here to shame me into doing it is useful.

**Comment by grumpyjames** [comment]
OK - had a look - we have perhaps three implementations internally, but they aren't heavily used, and, amusingly, they have implementation quirks that don't match up with the one provided here. I don't think we need to take any further action here.
